pkg/cri/resource-manager/policy/builtin/memtier/resources.go

```diff
@@ -496,38 +535,58 @@ func (cs *supply) Allocate(r Request) (Grant, error) {
 
 	cr := r.(*request)
 
-	// allocate isolated exclusive CPUs or slice them off the sharable set
-	switch {
-	case cr.full > 0 && cs.isolated.Size() >= cr.full && cr.isolate:
-		exclusive, err = cs.takeCPUs(&cs.isolated, nil, cr.full)
-		if err != nil {
-			return nil, policyError("internal error: "+
-				"%s: can't take %d exclusive isolated CPUs from %s: %v",
-				cs.node.Name(), cr.full, cs.isolated, err)
-		}
+	full := cr.full
+	fraction := cr.fraction
+	if cr.cpuType == cpuNormal {
+		// allocate isolated exclusive CPUs or slice them off the sharable set
+		switch {
+		case full > 0 && cs.isolated.Size() >= full && cr.isolate:
+			exclusive, err = cs.takeCPUs(&cs.isolated, nil, full)
+			if err != nil {
+				return nil, policyError("internal error: "+
+					"%s: can't take %d exclusive isolated CPUs from %s: %v",
+					cs.node.Name(), full, cs.isolated, err)
+			}
 
-	case cr.full > 0 && cs.AllocatableSharedCPU() > 1000*cr.full:
-		exclusive, err = cs.takeCPUs(&cs.sharable, nil, cr.full)
-		if err != nil {
+		case full > 0 && cs.AllocatableSharedCPU() > 1000*full:
+			exclusive, err = cs.takeCPUs(&cs.sharable, nil, full)
+			if err != nil {
+				return nil, policyError("internal error: "+
+					"%s: can't take %d exclusive CPUs from %s: %v",
+					cs.node.Name(), full, cs.sharable, err)
+			}
+
+		case full > 0:
 			return nil, policyError("internal error: "+
-				"%s: can't take %d exclusive CPUs from %s: %v",
-				cs.node.Name(), cr.full, cs.sharable, err)
+				"%s: can't slice %d exclusive CPUs from %s, %dm available",
+				cs.node.Name(), full, cs.sharable, cs.AllocatableSharedCPU())
+		}
+	} else if cr.cpuType == cpuReserved {
+		if full > 0 {
+			// no support for isolated allocation of Reserved CPUs
+			fraction += full * 1000
+			full = 0
 		}
```

Comment:
Despite of the earlier 'defensive programming' argument, I still don't like the idea of enforcing in two separate places that we have fractional-only allocations for containers in the `kube-system` namespace: once when we determine the container allocation preferences and once then when we do the allocation here.

If you really insist on doing it, then wouldn't it be much simpler to otherwise leave the original `switch` alone, but add this extra check/enforcement as a new first `case` there, in the form of:

```
    case full > 0 && cr.cpuType == cpuReserved:
        cr.fraction += cr.full + 1000
        cr.full = 0
```

Then instead of adding an if-else around the switch, changing its current indentation, touching every single case in it, and adding a branch for fractional allocation, you would be simply inserting a new (first) case in the switch and the necessary branching later for the fractional allocation.

Even splitting out the `cr.full > 0` test around the switch and adding the new case as `case cr.cpuType == cpuReserved:`
seems more natural to me than this... But for selfish reasons I'd prefer to go here with the minimal change set (I have a continuation-PR to #588 in the pipeline touching these same parts).

Reply:
And if we keep this extra check here, then a warning log message would be warranted IMO, too.

Reply:
Thanks @klihub! Warning added and indentation switched back to the original.

Unfortunately the switch-case still has some changes, because of using local full & fraction variables instead of attributes of the grant due to the warning and associated logic.